keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+- Labels
+- Nested parameters
+- Validation and ranged values
+- More parameter types (enums, numbers, etc.)
+- Easily extendable with custom attributes like `trigger` and `immutable`
+- Rich ecosystem of tools and libraries e.g. [UI generators](http://json-schema.org/implementations.html#generators-from-schemas)
+
+Latter point might help to build an ecosystem around KUDO operators which might help driving KUDOs adoption.
+
+### Goals
+
+Introduce a new parameter format based on the JSON-schema.
+
+### Non-Goals
+
+Extend the current format to support new features.
+
+### Proposal Overview
+
+One part of the proposal deals the client-side of the new format. This includes:
+ - Introduce new API version (`apiVersion:kudo.dev/v1beta2`) for the operator package `params.yaml`. KUDO should be able to read both old and new formats.
+- Support updating nested values from the command line e.g. `kudo update ... -p some.nested.key=value`
+- Support updating nested values using values file e.g. `kudo update ... -P values.yaml`
+- Add nested parameters via. the wizard `kudo package add parameter`
+- Parameter listing aka `kudo package list parameters`. We need a new output format here - maybe a tree-table?
+- Possible additional validation for our custom attributes e.g. parameter `trigger`
+
+The server-side of the proposal includes:
+- Updating `Operator` and `Instance` CRDs to use JSON-schema as the parameter format  
+- Providing a conversion-webhook to convert in-cluster running operators to the new format 
+   
+### Package Format
+
+JSON-schema is a [well-described](http://json-schema.org/understanding-json-schema/) format. However, here is a little example showing an existing `ZOOKEEPER_URI` and its new counterpart `Zookeeper.URI`:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: ZOOKEEPER_URI
+    displayName: "zookeeper cluster URI"
+    description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+    default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: true
+```
+
+```yaml
+type: object
+"$schema": "http://json-schema.org/draft-07/schema#"
+apiVersion: kudo.dev/v1beta2
+properties:
+  Zookeeper:
+    properties:
+      URI:
+        type: "string"
+        description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+        default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: ["URI"]
```

Comment:
the semantics of this required seems different to me...   the value of the required before was to ensure that a value for that property was provided... and was mainly useful when a default was NOT provided.   When a default was provided it was used.

Defining required in a schema feels different in that if a default value is available, the user does not need to provide it... but if you have a value file what you wanted to use the schema to validate it... it would fail because the required element is missing.

Reply:
That's a good point. I guess, we'll have to enrich the values with the defaults from the schema and then validate it 🤔

Reply:
Yeah - we have to merge the data before validating. There are discussions about this, and it seems to be consensus that validation will (usually) ignore default values.